src/Storages/MergeTree/ReplicatedMergeTreeAttachThread.cpp

```diff
@@ -0,0 +1,178 @@
+#include <Storages/MergeTree/ReplicatedMergeTreeAttachThread.h>
+#include <Storages/StorageReplicatedMergeTree.h>
+
+namespace DB
+{
+
+ReplicatedMergeTreeAttachThread::ReplicatedMergeTreeAttachThread(StorageReplicatedMergeTree & storage_)
```

Comment:
Maybe we can make it a part of `ReplicatedMergeTreeRestartingThread`? It does almost the same thing (trying to establish connection with ZooKeeper and startup table) with the only difference, that restarting threads omits checks that are done in `StorageReplicatedMergeTree` ctor. I think we can add a method that does this checks in `ReplicatedMergeTreeRestartingThread` and simply call it before `tryStartup`.

Reply:
Thought about it but I have a feeling it would only complicate it a lot. 
The biggest difference is that the `AttachThread` contains one-and-done operations as opposed to the `RestartingThread`. I don't see a reason why we would retry everything when a single thing fails. I can add a phase there and skip it if it's initialized, but then we would handle differently logic for checks and logic for restarting (retry chunks vs retry everything on each fail).
Also, we have `startup` function that has some interaction with attaching/initialization.

Based on all that, I think it's more readable this way, we have fire and forget with attaching and once it's done, the storage continues working normally.